server/src/main/java/org/apache/iotdb/db/engine/storagegroup/timeindex/DeviceTimeIndex.java

```diff
@@ -0,0 +1,294 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.db.engine.storagegroup.timeindex;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.iotdb.db.engine.StorageEngine;
+import org.apache.iotdb.db.exception.PartitionViolationException;
+import org.apache.iotdb.db.rescon.CachedStringPool;
+import org.apache.iotdb.db.utils.FilePathUtils;
+import org.apache.iotdb.db.utils.SerializeUtils;
+import org.apache.iotdb.tsfile.utils.RamUsageEstimator;
+import org.apache.iotdb.tsfile.utils.ReadWriteIOUtils;
+
+public class DeviceTimeIndex implements ITimeIndex {
+
+  protected static final int INIT_ARRAY_SIZE = 64;
+
+  protected static final Map<String, String> cachedDevicePool = CachedStringPool.getInstance()
+      .getCachedPool();
+
+  /**
+   * start times array.
+   */
+  protected long[] startTimes;
+
+  /**
+   * end times array. The values in this array are Long.MIN_VALUE if it's an unsealed sequence
+   * tsfile
+   */
+  protected long[] endTimes;
+
+  /**
+   * device -> index of start times array and end times array
+   */
+  protected Map<String, Integer> deviceToIndex;
+
+  public DeviceTimeIndex() {
+    init();
+  }
+
+  public DeviceTimeIndex(Map<String, Integer> deviceToIndex, long[] startTimes, long[] endTimes) {
+    this.startTimes = startTimes;
+    this.endTimes = endTimes;
+    this.deviceToIndex = deviceToIndex;
+  }
+
+  @Override
+  public void init() {
+    this.deviceToIndex = new ConcurrentHashMap<>();
+    this.startTimes = new long[INIT_ARRAY_SIZE];
+    this.endTimes = new long[INIT_ARRAY_SIZE];
+    initTimes(startTimes, Long.MAX_VALUE);
+    initTimes(endTimes, Long.MIN_VALUE);
+  }
+
+  @Override
+  public void serialize(OutputStream outputStream) throws IOException {
+    ReadWriteIOUtils.write(deviceToIndex.size(), outputStream);
+    for (Entry<String, Integer> entry : deviceToIndex.entrySet()) {
+      ReadWriteIOUtils.write(entry.getKey(), outputStream);
+      ReadWriteIOUtils.write(startTimes[entry.getValue()], outputStream);
+      ReadWriteIOUtils.write(endTimes[entry.getValue()], outputStream);
+    }
```

Comment:
How about serializing the three fields in a columnar style? This way we may be able to encode them.

Reply:
Very good point! Thanks!